rdtools/soiling.py

```diff
@@ -999,3 +1012,1287 @@ def monthly_soiling_rates(soiling_interval_summary, min_interval_length=14,
+                    change_point = ic
+                    if verbose:
+                        print('Now not assuming perfect cleaning')
+                elif (not perfect_cleaning
+                      and (ic >= max_iterations
+                           or (ic >= change_point + n_steps
+                               and relative_improvement
+                               < convergence_criterium))):
+                    if verbose:
+                        if relative_improvement < convergence_criterium:
+                            print('Convergence reached.')
+                        else:
+                            print('Max iterations reached.')
+                    ic = max_iterations
+
+        # Initialize output DataFrame
+        df_out = pd.DataFrame(index=pi.index,
+                              columns=['soiling_ratio', 'soiling_rates',
+                                       'cleaning_events', 'seasonal_component',
+                                       'degradation_trend', 'total_model',
+                                       'residuals'])
+
+        # Save values
+        df_out.seasonal_component = seasonal_component[-1]
+        df_out.degradation_trend = degradation_trend[-1]
+        degradation = yoy_save[-1]
+        final_kdf = soiling_dfs[-1]
+        df_out.soiling_ratio = final_kdf.soiling_ratio
+        df_out.soiling_rates = final_kdf.soiling_rates
+        df_out.cleaning_events = final_kdf.cleaning_events
+
+        # Calculate soiling loss in %
+        soiling_loss = (1 - df_out.soiling_ratio).mean() * 100
+
+        # Total model
+        df_out.total_model = (df_out.soiling_ratio
+                              * df_out.seasonal_component
+                              * df_out.degradation_trend)
+        df_out.residuals = pi / df_out.total_model
+        residual_shift = df_out.residuals.mean()
+        df_out.total_model *= residual_shift
+        RMSE = _RMSE(pi, df_out.total_model)
+        adf_res = adfuller(df_out.residuals.dropna(), regression='ctt', autolag=None)
+        if verbose:
+            print('p-value for the H0 that there is a unit root in the'
+                  + 'residuals (using the Augmented Dickey-fuller test):'
+                  + '{:.3e}'.format(adf_res[1]))
+
+        # Check size of soiling signal vs residuals
+        SR_amp = float(np.diff(df_out.soiling_ratio.quantile([.025, .975])))
+        residuals_amp = float(np.diff(df_out.residuals.quantile([.025, .975])))
+        soiling_signal_strength = SR_amp / residuals_amp
+        if soiling_signal_strength < soiling_significance_knob:
+            if verbose:
+                print('Soiling signal is small relative to the noise')
+            small_soiling_signal = True
+            df_out.SR_high = 1.0
+            df_out.SR_low = 1.0 - SR_amp
+
+        return df_out, degradation, soiling_loss, residual_shift, RMSE, \
+            small_soiling_signal, adf_res
+
+    def run_bootstrap(self, reps=512, verbose=False,
```

Comment:
It would be great to give `run_bootstap()` a `confidence_level` parameter so users can adjust the confidence level used for confidence intervals, similar to `degradation_year_on_year()` and `bootrap_YOY()`

Reply:
Good idea. I've included this in the coming commits.